vcsclient/github.go

```diff
@@ -0,0 +1,189 @@
+		if err != nil {
+			return nil, err
+		}
+		for _, repo := range repos {
+			results[*repo.Owner.Login] = append(results[*repo.Owner.Login], *repo.Name)
+		}
+		if nextPage+1 >= response.LastPage {
+			break
+		}
+	}
+	return results, nil
+}
+
+func (client *GitHubClient) ListBranches(owner, repository string) ([]string, error) {
+	branches, _, err := client.ghClient.Repositories.ListBranches(client.context, owner, repository, nil)
+	if err != nil {
+		return []string{}, err
+	}
+
+	results := []string{}
+	for _, repo := range branches {
+		results = append(results, *repo.Name)
+	}
+	return results, nil
+}
+
+func (client *GitHubClient) CreateWebhook(owner, repository, branch, payloadUrl string, webhookEvents ...vcsutils.WebhookEvent) (string, string, error) {
+	token := vcsutils.CreateToken()
+	hook := createGitHubHook(token, payloadUrl, webhookEvents...)
+	responseHook, _, err := client.ghClient.Repositories.CreateHook(client.context, owner, repository, hook)
+	if err != nil {
+		return "", "", err
+	}
+	return strconv.FormatInt(*responseHook.ID, 10), token, err
+}
+
+func (client *GitHubClient) UpdateWebhook(owner, repository, branch, payloadUrl, token, webhookId string, webhookEvents ...vcsutils.WebhookEvent) error {
+	webhookIdInt64, err := strconv.ParseInt(webhookId, 10, 64)
+	if err != nil {
+		return err
+	}
+	hook := createGitHubHook(token, payloadUrl, webhookEvents...)
+	_, _, err = client.ghClient.Repositories.EditHook(client.context, owner, repository, webhookIdInt64, hook)
+	return err
+}
+
+func (client *GitHubClient) DeleteWebhook(owner, repository, webhookId string) error {
+	webhookIdInt64, err := strconv.ParseInt(webhookId, 10, 64)
+	if err != nil {
+		return err
+	}
+	_, err = client.ghClient.Repositories.DeleteHook(client.context, owner, repository, webhookIdInt64)
+	return err
+}
+
+func (client *GitHubClient) SetCommitStatus(commitStatus CommitStatus, owner, repository, ref, title, description, detailsUrl string) error {
+	state := getGitHubCommitState(commitStatus)
+	status := &github.RepoStatus{
+		Context:     &title,
+		TargetURL:   &detailsUrl,
+		State:       &state,
+		Description: &description,
+	}
+	_, _, err := client.ghClient.Repositories.CreateStatus(client.context, owner, repository, ref, status)
+	return err
+}
+
+func (client *GitHubClient) DownloadRepository(owner, repository, branch, localPath string) error {
+	url, _, err := client.ghClient.Repositories.GetArchiveLink(client.context, owner, repository, github.Tarball, &github.RepositoryContentGetOptions{Ref: branch}, true)
+	if err != nil {
+		return err
+	}
+
+	httpClient := &http.Client{}
+	req, err := http.NewRequest("GET", url.String(), nil)
+	req.Header.Add("Accept", "application/vnd.github.v3+json")
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	return vcsutils.Untar(localPath, resp.Body, true)
+}
+
+func (client GitHubClient) Push(owner, repository string, branch string) error {
```

Comment:
This function should probably be deleted. The same goes for the other providers of course.